satellite/gracefulexit/endpoint_test.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gracefulexit_test
+
+import (
+	"io"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/internal/testrand"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storagenode"
+	"storj.io/storj/uplink"
+)
+
+const numMessages = 6
+
+func TestSuccess(t *testing.T) {
+	testTransfers(t, numMessages, func(ctx *testcontext.Context, satellite *testplanet.SatelliteSystem, processClient pb.SatelliteGracefulExit_ProcessClient, exitingNode *storagenode.Peer) {
+		var pieceID storj.PieceID
+		for {
+			response, err := processClient.Recv()
+			if err == io.EOF {
+				// Done
+				break
+			}
+			require.NoError(t, err)
+
+			switch m := response.GetMessage().(type) {
+			case *pb.SatelliteMessage_TransferPiece:
+				require.NotNil(t, m)
+
+				// pick the first one to fail
+				if pieceID.IsZero() {
+					pieceID = m.TransferPiece.PieceId
+				}
+
+				if pieceID != m.TransferPiece.PieceId {
+					success := &pb.StorageNodeMessage{
+						Message: &pb.StorageNodeMessage_Succeeded{
+							Succeeded: &pb.TransferSucceeded{
+								OriginalPieceHash: &pb.PieceHash{PieceId: m.TransferPiece.PieceId},
+								AddressedOrderLimit: &pb.AddressedOrderLimit{
+									Limit: &pb.OrderLimit{
+										PieceId: m.TransferPiece.PieceId,
+									},
+								},
+							},
+						},
+					}
+					err = processClient.Send(success)
+					require.NoError(t, err)
+				} else {
+					failed := &pb.StorageNodeMessage{
+						Message: &pb.StorageNodeMessage_Failed{
+							Failed: &pb.TransferFailed{
+								PieceId: m.TransferPiece.PieceId,
+								Error:   pb.TransferFailed_UNKNOWN,
+							},
+						},
+					}
+					err = processClient.Send(failed)
+					require.NoError(t, err)
+				}
+			case *pb.SatelliteMessage_ExitCompleted:
+				// TODO test completed signature stuff
+				break
+			default:
+				t.FailNow()
+			}
+		}
+
+		// check that the exit has completed and we have the correct transferred/failed values
+		progress, err := satellite.DB.GracefulExit().GetProgress(ctx, exitingNode.ID())
+		require.NoError(t, err)
+
+		require.Equal(t, int64(numMessages), progress.PiecesTransferred)
+		// even though we failed 1, it eventually succeeded, so the count should be 0
```

Comment:
I don't see how the first piece eventually succeeded to transfer. It seems like the switch statement above will always fail the first piece we attempt to transfer.

Reply:
The pieceID changes every time an order limit is created for a given piece.
So this sets it once and the check after the first failure has a different pieceID.
```
if pieceID.IsZero() {
	pieceID = m.TransferPiece.PieceId
}
```
Should this not be?

Reply:
fixed